source/common/stats/statsd.cc

```diff
@@ -53,24 +54,52 @@ void Writer::send(const std::string& message) {
 }
 
 UdpStatsdSink::UdpStatsdSink(ThreadLocal::SlotAllocator& tls,
-                             Network::Address::InstanceConstSharedPtr address)
-    : tls_(tls.allocateSlot()), server_address_(address) {
+                             Network::Address::InstanceConstSharedPtr address, const bool use_tag)
+    : tls_(tls.allocateSlot()), server_address_(address), use_tag_(use_tag) {
   tls_->set([this](Event::Dispatcher&) -> ThreadLocal::ThreadLocalObjectSharedPtr {
     return std::make_shared<Writer>(this->server_address_);
   });
 }
 
 void UdpStatsdSink::flushCounter(const Counter& counter, uint64_t delta) {
-  tls_->getTyped<Writer>().writeCounter(counter.name(), delta);
+  tls_->getTyped<Writer>().writeCounter(getName(counter), buildTagStr(counter.tags()), delta);
 }
 
 void UdpStatsdSink::flushGauge(const Gauge& gauge, uint64_t value) {
-  tls_->getTyped<Writer>().writeGauge(gauge.name(), value);
+  tls_->getTyped<Writer>().writeGauge(getName(gauge), buildTagStr(gauge.tags()), value);
 }
 
 void UdpStatsdSink::onHistogramComplete(const Histogram& histogram, uint64_t value) {
   // For statsd histograms are all timers.
-  tls_->getTyped<Writer>().writeTimer(histogram.name(), std::chrono::milliseconds(value));
+  tls_->getTyped<Writer>().writeTimer(getName(histogram), buildTagStr(histogram.tags()),
+                                      std::chrono::milliseconds(value));
+}
+
+std::string UdpStatsdSink::getName(const Metric& metric) {
+  if (use_tag_) {
+    return metric.tagExtractedName();
+  } else {
+    return metric.name();
+  }
+}
+
+const std::string UdpStatsdSink::buildTagStr(const std::vector<Tag>& tags) {
+  if (!use_tag_ || tags.empty()) {
+    return {};
+  }
+
+  std::string tag_str = "|#";
+  int i = 0;
+  for (const Tag& tag : tags) {
+    if (i == 0) {
+      std::vector<std::string> v = {tag.name_, tag.value_};
+      tag_str.append(StringUtil::join(v, ":"));
```

Comment:
The idea with join is you do something like:
```
std::vector<std::string> tag_strs;
for (const Tag& tag : tags) {
  tag_strs.emplace_back(tag._name_ + ":" + tag._value);
}
return "|#" + StringUtil::join(tag_strs, ",");
```

This is probably less performant than what you have, but more concise. Since this is only at statsd flush time, it's probably fine; if we hit scaling issues then we can micro-optimize.

Reply:
>Since this is only at statsd flush time, it's probably fine

Agreed. Refine with 3c386683bf5c102b4d9dc614805d63b0a843450a.